### PR TITLE
feat: on call notification channel separation

### PR DESCRIPTION
**Pull request by andrewbenington**
- [x] Identified the issue which this PR solves.
- [x] Read the [**CONTRIBUTING**](https://github.com/target/goalert/blob/master/CONTRIBUTING.md) document.
- [x] Code builds clean without any errors or warnings.
- [x] Added appropriate tests for any new functionality.
- [x] All new and existing tests passed.
- [x] Added comments in the code, where necessary.
- [x] Ran `make check` to catch common errors. Fixed any that came up.

**Description:**

- The dialog for creating/editing schedule on-call notifications now features a dropdown to select channel type
- Slack channel/user groups have been separated into two channel types, selectable from the dropdown
  - The checkbox was removed, as its functionality is replaced by the dropdown
- User Group notifications are no longer sent to the Slack channel unless there is an error with the User Group
- The `Value` type for `ScheduleOnCallNotificationsForm` has been altered to be more accommodating of different notification channel types
- Minor dev fix: use `YARN_VERSION` instead of assumed stable version
  - Fixes `Usage Error: Invalid package manager specification in CLI arguments; expected a semver version` from recent build changes

**Which issue(s) this PR fixes:**
#3045 

**Screenshots:**
If applicable, add some screenshots here.
Screenshot 2023-05-24 at 4 21 43 PM
Screenshot 2023-05-24 at 4 21 47 PM

image

**Describe any introduced user-facing changes:**
To use Slack User Groups for schedule on-call notifications, the user must select "SLACK USER GROUP" from the new dropdown instead of checking a checkbox.

### CODE REVIEW

**Review by mastercactapus — CHANGES_REQUESTED** [review]
For the layout:
- can we move the notification type selection below the notification options?
- The Slack channel for user groups is for errors/issues, not as a fallback (from a user perspective)

Also, this needs the behavioral change in the backend -- ensuring user group updates don't send the regular channel notification.

**Comment by andrewbenington** [comment]
> For the layout:
> 
>     * can we move the notification type selection below the notification options?
> 
>     * The Slack channel for user groups is for errors/issues, not as a fallback (from a user perspective)
> 
> 
> Also, this needs the behavioral change in the backend -- ensuring user group updates don't send the regular channel notification.

Just to clarify - no messages should be sent at all if there isn't an error updating the UG?

**Comment by mastercactapus** [comment]
Correct, if the notification channel is `SLACK_USER_GROUP` no messages should be sent to the channel unless there is an error.

File: web/src/app/schedules/on-call-notifications/util.ts [hunk 1]
```diff
@@ -4,20 +4,115 @@ import { DateTime } from 'luxon'
...
+  return target
+}
+
 export type Value = {
-  slackChannelID?: string | null
-  slackUserGroup?: string | null
   time: string | null
   weekdayFilter: WeekdayFilter
+  type: NotificationChannelType
+  channelFields: ChannelFields
```
Comment (andrewbenington) [inline web/src/app/schedules/on-call-notifications/util.ts:105]: Future-proofing for non-slack channels

File: notification/slack/usergrouperror.go [hunk 2]
```diff
@@ -19,7 +19,7 @@ type userGroupError struct {
 	ScheduleName string
 	Missing      []notification.User
 
-	OnCall string
+	ErrorMsg string
```
Comment (mastercactapus) [inline notification/slack/usergrouperror.go:22]: This field and the `{{.ErrorMsg}}` line can be removed entirely, that was only used as the placeholder for the on call message. The entire template _is_ the error message

File: notification/slack/usergroupsender.go [hunk 3]
```diff
@@ -132,27 +131,31 @@ func (s *UserGroupSender) Send(ctx context.Context, msg notification.Message) (*
...
+	// Only send to the channel if an error occurred
+	if stateDetails != "" {
+		err = s.withClient(ctx, func(c *slack.Client) error {
+			_, ts, err = c.PostMessageContext(ctx, chanID, slack.MsgOptionText(errorMsg, false))
+			if err != nil {
+				return fmt.Errorf("post message to channel '%s': %w", chanID, err)
+			}
+			return nil
+		})
 		if err != nil {
```
Comment (mastercactapus) [inline notification/slack/usergroupsender.go:156]: Let's flatten this.
https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

We can invert the check for the quick/easy return (since there's no state details or external ID). Secondly, we can just return the error to `withClient` and add the context in the external error check.
```go
if stateDetails == "" {
	return &notification.SentMessage{State: notification.StateDelivered}, nil
}

err = s.withClient(ctx, //...
	_, ts, err = c.PostMessageContext( //...
	return err
})
if err != nil {
	return nil, fmt.Errorf("post message to channel '%s': %w", chanID, err)
}
```

**Comment by Forfold** [comment]
We can probably disable the type field if there is only 1 type present/user groups are disabled:

Screenshot 2023-05-31 at 12 16 29 PM

**Comment by andrewbenington** [comment]
> We can probably disable the type field if there is only 1 type present/user groups are disabled:
> 
> Also maybe make the types Title Case?
> 
> Screenshot 2023-05-31 at 12 16 29 PM

Good idea about the type field.

And I made the types in all caps to stay in line with the personal contact method dropdown, which uses all caps. I'm fine with changing that to title case as well, but I think we should be consistent

**Comment by Forfold** [comment]
Good point, could probably be a small change in its own PR since it's on the profile too.

When I go to create a duplicate user group, it only validates that the channel is the same (even though it's for error reporting). I think we should error on a duplicate user-group and set that as the first error to return if using that type, even if the channel for error messaging is different.

Screenshot 2023-05-31 at 12 27 12 PM
Screenshot 2023-05-31 at 12 27 28 PM

**Comment by andrewbenington** [comment]
> Good point, could probably be a small change in its own PR since it's on the profile too.
> 
> When I go to create a duplicate user group, it only validates that the channel is the same (even though it's for error reporting). I think we should error on a duplicate user-group and set that as the first error to return if using that type, even if the channel for error messaging is different. This could also be a separate issue I think but we could maybe fix here.
> 
> Screenshot 2023-05-31 at 12 27 12 PM Screenshot 2023-05-31 at 12 27 28 PM

It will only show the error if both the user group and channel are the same as another existing rule. @mastercactapus says he thinks that's fine for now, and that we can refactor this in the future if we want to make the user group unique between rules.

File: web/src/app/schedules/on-call-notifications/ScheduleOnCallNotificationsEditDialog.tsx [hunk 4]
```diff
@@ -32,7 +26,7 @@ export default function ScheduleOnCallNotificationsEditDialog(
       ? DateTime.fromFormat(rule.time, 'HH:mm', { zone }).toISO()
       : null,
     weekdayFilter: rule?.time ? rule.weekdayFilter || EVERY_DAY : NO_DAY,
-    type: channelTypeFromTarget(rule?.target),
+    type: rule?.target?.type ?? 'slackChannel',
```
Comment (Forfold) [inline web/src/app/schedules/on-call-notifications/ScheduleOnCallNotificationsEditDialog.tsx:29]: Does the type of slack channel no longer come in? If there's no type it seems odd to assume slackChannel

I feel like this may produce an issue in the future if we add more types- the code before was explicit in what types come in using a switch statement

File: Makefile [hunk 5]
```diff
@@ -181,7 +181,7 @@ ensure-yarn: # Yarn ensures the correct version of yarn is installed
 
 yarn:
 	corepack enable
-	corepack prepare yarn@stable --activate
+	corepack prepare yarn@$(YARN_VERSION) --activate
```
Comment (Forfold) [inline Makefile:184]: Is this change intended? It doesn't seem applicable to this PR

Comment (andrewbenington) [reply Makefile:184]: `make yarn` fails with `Usage Error: Invalid package manager specification in CLI arguments; expected a semver version` otherwise. I figured I would include it as a minor dev fix, but if you think it would be better as its own PR then I can revert it.

Comment (Forfold) [reply Makefile:184]: Sounds good. I ran into this too at some point. I updated the description.

**Review by Forfold — APPROVED** [review]
